### PR TITLE
vcsim: add PropertyCollector index support

### PR DESCRIPTION
**Pull request by dougm**
PropertyCollector supports use of index (unique key) to collect a single array element. Further, a property path may also be applied to the given element.

### CODE REVIEW

File: govc/test/object.bats [hunk 1]
```diff
@@ -411,6 +411,66 @@ EOF
...
+  run govc object.collect -o $GOVC_VM 'config["guestinfo.a"]'
+  assert_failure
+
+  run govc object.collect -o $GOVC_VM 'config.extraConfig["guestinfo.a"]'
+  assert_success # Key does not exist, not an error
+
+  run govc vm.change -e "guestinfo.a=1" -e "guestinfo.b=2"
+  assert_success
+
+  run govc object.collect -json $GOVC_VM 'config.extraConfig["guestinfo.b"]'
```
Comment (hickeng) [inline govc/test/object.bats:458]: If you have `guestinfo.abc=3` does that match (ie. substring match) or is it doing an exact token match between the separators?

Comment (dougm) [reply govc/test/object.bats:458]: It's exact afaik, equal to the `Key` field of the array type.

File: simulator/property_collector.go [hunk 2]
```diff
@@ -214,6 +218,64 @@ func fieldValue(rval reflect.Value, p string) (interface{}, error) {
...
+			item = item.Elem()
+		}
+		if item.Kind() == reflect.Ptr {
+			item = item.Elem()
+		}
+		if item.Kind() != reflect.Struct {
+			return reflect.Value{}, errInvalidField
+		}
+
+		field := item.FieldByName("Key")
```
Comment (akutz) [inline simulator/property_collector.go:240]: I am having trouble tracing how we get here, so maybe this is a non-issue, but what if the property path is for an array of non-struct types, such as an array of integers or strings? I believe a numerical index should work in that case.

Comment (dougm) [reply simulator/property_collector.go:240]: afaik, property path array index can only be used when struct has a `key` field, see for example:

```console
% govc object.collect $GOVC_VM disabledMethod
disabledMethod  []string  MakePrimaryVM_Task,TerminateFaultTolerantVM_Task,...

% govc object.collect $GOVC_VM 'disabledMethod[1]'
govc: ServerFaultCode: InvalidProperty
```

Another case, AuthorizationRole has no `key` field. And while it does have a unique `roleId` field, same `InvalidProperty` result when using a valid `roleId`
```console
% id=$(govc object.collect -o -json AuthorizationManager:AuthorizationManager roleList | jq .roleList[].roleId | head -1)

% govc object.collect AuthorizationManager:AuthorizationManager "roleList[$id]"
govc: ServerFaultCode: InvalidProperty
```

This [doc](https://docs.vmware.com/en/VMware-vSphere/8.0/vsphere-web-services-programming-guide/GUID-DB836CA6-8316-4B2F-8CE5-8AD933F8DE36.html) also seems to indicate must be key-based:

> Nested properties can also refer to entries in a key-based array

**Review by aruneshpa — APPROVED** [review]

**Pull request merged**